### PR TITLE
fix showing multiple failure matches per template on -ms set

### PR DESCRIPTION
**Pull request by RamanaReddy0M**
#2693 Bug in matcher-status option

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/pkg/protocols/common/executer/executer.go [hunk 1]
```diff
@@ -15,6 +15,10 @@ import (
 	"github.com/projectdiscovery/nuclei/v2/pkg/protocols/common/helpers/writer"
 )
 
+// ShowFailureMatchPerRequest if set to true, will show failure match per request
+// by defalut shows failure match per template.
+var ShowFailureMatchPerRequest bool
+
 // Executer executes a group of requests for a protocol
```
Comment (tarunKoyalwar) [inline v2/pkg/protocols/common/executer/executer.go:22]: @RamanaReddy0M requesting minor change regarding implementation . we can move this variable to types.Options and access it here in Executor with `e.options.Options.MatchStatusPerRequest`

also this variable seems to  be unused and hence always false

- [ ] add new flag `-msr -matcher-status-request` to show matcher status per request (use ^ variable)

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]
requesting minor changes + add new flag to control behaviour of matcher status at template scope vs per request

File: v2/cmd/integration-test/http.go [hunk 2]
```diff
@@ -78,6 +78,7 @@ var httpTestcases = map[string]testutils.TestCase{
 	"http/cl-body-with-header.yaml":                 &httpCLBodyWithHeader{},
 	"http/save-extractor-values-to-file.yaml":       &httpSaveExtractorValuesToFile{},
 	"http/cli-with-constants.yaml":                  &ConstantWithCliVar{},
+	"http/matcher-status.yaml":                      &matcherStatusTest{},
 }
```
Comment (tarunKoyalwar) [inline v2/cmd/integration-test/http.go:82]: add new integration test along with this one where we use `-msr` flag

File: v2/pkg/output/output.go [hunk 3]
```diff
@@ -302,7 +304,7 @@ func (w *StandardWriter) Close() {
 
 // WriteFailure writes the failure event for template to file and/or screen.
 func (w *StandardWriter) WriteFailure(event InternalEvent) error {
-	if !w.matcherStatus {
+	if !w.matcherStatus && !w.matchStatusPerRequest {
 		return nil
```
Comment (tarunKoyalwar) [inline v2/pkg/output/output.go:308]: @RamanaReddy0M  is it possible to handle all logic related to matchers status per request in common/executer/executer.go . without updating  writer methods ??

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]
`-msr` should implicitly enable `-ms` to avoid conflicting scenarios
```
$ ./nuclei -u example.com -id CVE-2021-29490  -msr   

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.9.6

		projectdiscovery.io

[WRN] Found 3 templates loaded with deprecated protocol syntax, update before v2.9.5 for continued support.
[INF] Current nuclei version: v2.9.6 (latest)
[INF] Current nuclei-templates version: v9.5.2 (latest)
[INF] New templates added in latest release: 50
[INF] Templates loaded for current scan: 1
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[CVE-2021-29490] [http] [medium] https://example.com
[CVE-2021-29490] [http] [medium] https://example.com
[INF] No results found. Better luck next time!
```

^ output shows that target is vulnerable but actually i.e matcher status (and status is omitted)

- suggesting updating PR description with POC . ex: command to run to test implementation

**Comment by RamanaReddy0M** [comment]
- matcher-status
```console
✗ ./nuclei -u example.com -id CVE-2021-29490  -ms 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.9.6

                projectdiscovery.io

[INF] Current nuclei version: v2.9.6 (latest)
[INF] Current nuclei-templates version: v9.5.3 (latest)
[INF] New templates added in latest release: 82
[INF] Templates loaded for current scan: 1
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[CVE-2021-29490] [failed] [http] [medium] https://example.com
```

- matcher status per request
```console
✗ ./nuclei -u example.com -id CVE-2021-29490  -msr

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.9.6

                projectdiscovery.io

[INF] Current nuclei version: v2.9.6 (latest)
[INF] Current nuclei-templates version: v9.5.3 (latest)
[INF] New templates added in latest release: 82
[INF] Templates loaded for current scan: 1
[INF] Targets loaded for current scan: 1
[INF] Running httpx on input host
[INF] Found 1 URL from httpx
[CVE-2021-29490] [failed] [http] [medium] https://example.com
[CVE-2021-29490] [failed] [http] [medium] https://example.com
```

**Review by tarunKoyalwar — APPROVED** [review]
lgtm !

## when using `-msr` (matcher status per request)
```console
$ echo https://example.com | ./nuclei -tags oast -msr -silent | sort | uniq -c | sort -r
   7 [CVE-2023-27350] [failed] [http] [critical] https://example.com
   5 [request-based-interaction] [failed] [http] [info] https://example.com
   4 [CVE-2022-22965] [failed] [http] [critical] https://example.com
   3 [umbraco-base-ssrf] [failed] [http] [medium] https://example.com
   3 [fastjson-1-2-68-rce] [failed] [http] [critical] https://example.com
   3 [CVE-2022-22947] [failed] [http] [critical] https://example.com
   3 [CVE-2022-21587] [failed] [http] [critical] https://example.com
   3 [CVE-2020-28976] [failed] [http] [medium] https://example.com
   2 [fastjson-1-2-24-rce] [failed] [http] [critical] https://example.com
   2 [apache-solr-log4j-rce] [failed] [http] [critical] https://example.com
   2 [CVE-2022-31814] [failed] [http] [critical] https://example.com
   2 [CVE-2022-26134] [failed] [http] [critical] https://example.com
   2 [CVE-2022-24112] [failed] [http] [critical] https://example.com
   2 [CVE-2022-22972] [failed] [http] [critical] https://example.com
   2 [CVE-2021-44228] [failed] [http] [critical] https://example.com
   2 [CVE-2021-29490] [failed] [http] [medium] https://example.com
   2 [CVE-2021-21881] [failed] [http] [critical] https://example.com
   2 [CVE-2021-1497] [failed] [http] [critical] https://example.com
   2 [CVE-2020-28188] [failed] [http] [critical] https://example.com
   2 [CVE-2020-25506] [failed] [http] [critical] https://example.com
   2 [CVE-2020-17456] [failed] [http] [critical] https://example.com
   2 [CVE-2019-20224] [failed] [http] [high] https://example.com
   2 [CVE-2019-12988] [failed] [http] [critical] https://example.com
   2 [CVE-2019-12987] [failed] [http] [critical] https://example.com
   2 [CVE-2019-12986] [failed] [http] [critical] https://example.com
   2 [CVE-2019-12985] [failed] [http] [critical] https://example.com
   2 [CVE-2018-10818] [failed] [http] [critical] https://example.com
   2 [CVE-2018-10562] [failed] [http] [critical] https://example.com
   2 [CVE-2017-10271] [failed] [http] [high] https://example.com
   1 [zimbra-preauth-ssrf] [failed] [http] [critical] https://example.com
   1 [xenmobile-server-log4j] [failed] [http] [critical] https://example.com
   1 [wp-xmlrpc-pingback-detection] [failed] [http] [info] https://example.com
   1 [wordpress-ssrf-oembed] [failed] [http] [medium] https://example.com
```

## when using `-ms` match (matcher status)
```console
$  echo https://example.com | ./nuclei -tags oast -ms -silent | sort | uniq -c | sort -r
   1 [zimbra-preauth-ssrf] [failed] [http] [critical] https://example.com
   1 [xenmobile-server-log4j] [failed] [http] [critical] https://example.com
   1 [wp-xmlrpc-pingback-detection] [failed] [http] [info] https://example.com
   1 [wordpress-ssrf-oembed] [failed] [http] [medium] https://example.com
   1 [webpagetest-ssrf] [failed] [http] [high] https://example.com
   1 [vrealize-operations-log4j-rce] [failed] [http] [critical] https://example.com
   1 [vmware-vcenter-ssrf] [failed] [http] [critical] https://example.com
   1 [vmware-vcenter-log4j-jndi-rce] [failed] [http] [critical] https://example.com
   1 [vmware-siterecovery-log4j-rce] [failed] [http] [critical] https://example.com
   1 [vmware-operation-manager-log4j] [failed] [http] [critical] https://example.com
   1 [vmware-nsx-log4j] [failed] [http] [critical] https://example.com
   1 [vmware-horizon-log4j-jndi-rce] [failed] [http] [critical] https://example.com
   1 [vmware-hcx-log4j] [failed] [http] [critical] https://example.com
   1 [unifi-network-log4j-rce] [failed] [http] [critical] https://example.com
   1 [unauth-ztp-ping] [failed] [http] [high] https://example.com
   1 [umbraco-base-ssrf] [failed] [http] [medium] https://example.com
   1 [tls-sni-proxy] [failed] [http] [info] https://example.com
   1 [targa-camera-ssrf] [failed] [http] [high] https://example.com
   1 [ssrf-via-oauth-misconfig] [failed] [http] [medium] https://example.com
   1 [springboot-log4j-rce] [failed] [http] [critical] https://example.com
```

### Note:
`-msr` does not show matchers status of each matcher ❌ 
it shows matcher status if there are multiple requests 👍

**Review by ehsandeep — CHANGES_REQUESTED** [review]
```console
./nuclei -pt tcp -tags cve -ms -u 1.1.1.1
```

- [x] not working with network protocol

**Comment by RamanaReddy0M** [comment]
```console
✗ ./nuclei -pt tcp -tags cve -ms -u 1.1.1.1 -silent
[CVE-2022-31793] [failed] [tcp] [high] 1.1.1.1
[CVE-2018-2628] [failed] [tcp] [critical] 1.1.1.1
[CVE-2016-2004] [failed] [tcp] [critical] 1.1.1.1
[CVE-2022-24706] [failed] [tcp] [critical] 1.1.1.1
[CVE-2015-3306] [failed] [tcp] [high] 1.1.1.1
[CVE-2017-3881] [failed] [tcp] [critical] 1.1.1.1
[CVE-2023-33246] [failed] [tcp] [critical] 1.1.1.1
[CVE-2020-1938] [failed] [tcp] [critical] 1.1.1.1
[CVE-2016-3510] [failed] [tcp] [critical] 1.1.1.1
[CVE-2021-44521] [failed] [tcp] [critical] 1.1.1.1
[CVE-2001-1473] [failed] [tcp] [high] 1.1.1.1
[CVE-2020-7247] [failed] [tcp] [critical] 1.1.1.1
[CVE-2018-2893] [failed] [tcp] [critical] 1.1.1.1
[CVE-2022-0543] [failed] [tcp] [critical] 1.1.1.1
[CVE-2011-2523] [failed] [tcp] [critical] 1.1.1.1
```

**Review by ehsandeep — CHANGES_REQUESTED** [review]
`-ms` option working as expected, we can remove `-msr` option as this feature is focused on template level match / fail status only.

-  [x] remove `-msr` option

Note: for later, `-ms` option does not consider the template failed to execute because of missing dynamic variable as failed.

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**